elements/research/ddiotune.cc

```diff
@@ -0,0 +1,385 @@
+{
+	uint32_t val;
+	if(!pacc)
+		init_pci_access();
+
+	struct pci_dev* dev=find_ddio_device(nic_bus);
+	if(!dev){
+		printf("No device found!\n");
+		printf("No device found!\n");
+		exit(1);
+	}
+	/*
+	 * perfctrlsts_0
+	 * bit 3: NoSnoopOpWrEn -> Should be 1b
+	 * bit 7: Use_Allocating_Flow_Wr -> Should be 0b
+	 * Check p. 68 of Intel® Xeon® Processor Scalable Family
+	 * Datasheet, Volume Two: Registers
+	 * May 2019
+	 * link: https://www.intel.com/content/www/us/en/processors/xeon/scalable/xeon-scalable-datasheet-vol-2.html
+	 * Haswell link: https://www.intel.com/content/dam/www/public/us/en/documents/datasheets/xeon-e5-v2-datasheet-vol-2.pdf
+	 */
+	val=pci_read_long(dev,PERFCTRLSTS_0);
+	printf("perfctrlsts_0 val: 0x%" PRIx32 "\n",val);
+	printf("NoSnoopOpWrEn val: 0x%" PRIx32 "\n",val&nosnoopopwren_MASK);
+	printf("Use_Allocating_Flow_Wr val: 0x%" PRIx32 "\n",val&use_allocating_flow_wr_MASK);
+	if(val&use_allocating_flow_wr_MASK)
+		return 1;
+	else
+		return 0;
+}
+
+/* Enable DDIO for the input NIC */
+void
+DDIOTune::ddio_enable(uint8_t nic_bus)
+{
+	uint32_t val;
+	if(!pacc)
+		init_pci_access();
+
+	if(!ddio_status(nic_bus))
+	{
+		struct pci_dev* dev=find_ddio_device(nic_bus);
+		if(!dev){
+			printf("No device found!\n");
+			exit(1);
+		}
+		val=pci_read_long(dev,PERFCTRLSTS_0);
+		pci_write_long(dev,PERFCTRLSTS_0,val|use_allocating_flow_wr_MASK);
+		printf("DDIO is enabled!\n");
+	} else
+	{
+		printf("DDIO was already enabled!\n");
+	}
+}
+
+/* Disable DDIO for the input NIC */
+void
+DDIOTune::ddio_disable(uint8_t nic_bus)
+{
+	uint32_t val;
+	if(!pacc)
+		init_pci_access();
+
+	if(ddio_status(nic_bus))
+	{
+		struct pci_dev* dev=find_ddio_device(nic_bus);
+		if(!dev){
+			printf("No device found!\n");
+			exit(1);
+		}
+		val=pci_read_long(dev,PERFCTRLSTS_0);
+		pci_write_long(dev,PERFCTRLSTS_0,val&(~use_allocating_flow_wr_MASK));
+		printf("DDIO is disabled!\n");
+	} else 
+	{
+		printf("DDIO was already disabled\n");
+	}
+}
+
+/* Print PCI root information */
+void
+DDIOTune::print_dev_info(struct pci_dev *dev)
+{
+	if(!dev){
+		printf("No device found!\n");
+		exit(1);
+	}
+	unsigned int c;
+	char namebuf[1024], *name;
+	printf("========================\n");
+	printf("%04x:%02x:%02x.%d vendor=%04x device=%04x class=%04x irq=%d (pin %d) base0=%lx \n",
+		dev->domain, dev->bus, dev->dev, dev->func, dev->vendor_id, dev->device_id,
+		dev->device_class, dev->irq, c, (long) dev->base_addr[0]);
+	name = pci_lookup_name(pacc, namebuf, sizeof(namebuf), PCI_LOOKUP_DEVICE, dev->vendor_id, dev->device_id);
+	printf(" (%s)\n", name);
+	printf("========================\n");
+}
+
+/* Constructor */
+DDIOTune::DDIOTune() : _w(-1)
+{
+	eax=0;ebx=0;ecx=0;edx=0;
+	_numa_node=-1;
+	_msr_cpu=-1;
+	IIO_LLC_WAYS_VALUE_NEW=0;
+	ways = new uint32_t[CACHE_LEVELS];
+	partitions = new uint32_t[CACHE_LEVELS];
+	linesize = new uint32_t[CACHE_LEVELS];
+	sets = new uint32_t[CACHE_LEVELS];
+	size = new uint32_t[CACHE_LEVELS];
+
+	init_pci_access();
```

Comment:
this should go to the initialize() method of the element

Reply:
I will move it to the configure method. Would it be okay?